drivers/apc_ul_util.h

```diff
@@ -0,0 +1,199 @@
+/*
+ * apc_ul_util.h - APC Microlink driver defines for descriptor parser and
+ *                 useful functions
+ *
+ *  Copyright (C)
+ *    2024 Thomas Eldon <tom@tomeldon.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#ifndef NUT_APC_UL_UTIL_H_SEEN
+#define NUT_APC_UL_UTIL_H_SEEN 1
+
+#include "apc_ul.h"
+#include "main.h"
+
+#define MAX_USAGES 512
+
+#define MAX_FQIDS 8
+
+#define FQUN_AS_STRING_MAX_LENGTH (13 * MAX_FQIDS) + 1
+#define USAGE_AS_STRING_MAX_LENGTH (FQUN_AS_STRING_MAX_LENGTH + 63)
+
+#define COLLECTION_IGNORE_LENGTH -2
+
+#define FQID(id_i, type_i) {.id = (id_i), .type = (type_i), .array_index = -1, .array_length = -1}
+#define FQID_INDEX(id_i, type_i, index_i, length_i) {.id = (id_i), .type = (type_i), .array_index = (index_i), .array_length = (length_i)}
+
+#define FQUN(...) \
+	{__VA_ARGS__}
+
+enum fqid_type
+{
+	fqid_namespace = 0,
+	fqid_collection = 1,
+	fqid_usage = 2
+};
+
+typedef struct
+{
+	uint8_t id;
+	enum fqid_type type;
+	/* Array details, negative length used for undetermined size */
+	int array_index;
+	int array_length;
```

Comment:
Why not e.g. `ssize_t` if we want signed (and to fit into compiler-guaranteed array sizing)?